enhancers.go

```diff
@@ -2,29 +2,325 @@ package main
-	involvedObject := fmt.Sprintf("%s/%s", objectRef.Kind, objectRef.Name)
+const breadcrumbLimit = 20
+
+func runEnhancers(ctx context.Context, eventObject *v1.Event, kind string, object metav1.Object, scope *sentry.Scope, sentryEvent *sentry.Event) error {
+
+	involvedObject := fmt.Sprintf("%s/%s", kind, object.GetName())
 	ctx, logger := getLoggerWithTag(ctx, "object", involvedObject)
+	logger.Debug().Msgf("Running the enhancer")
 
 	var err error
-	logger.Debug().Msgf("Running enhancers...")
 
-	// First, run the basic (common) enhancer
-	runCommonEnhancer(ctx, scope, sentryEvent)
+	// First, run the common enhancer
+	err = runCommonEnhancer(ctx, scope, sentryEvent)
+	if err != nil {
+		return err
+	}
+	// If an event object is provided, we call the event enhancer
+	if eventObject != nil {
+		err = eventEnhancer(ctx, scope, eventObject, sentryEvent)
+		if err != nil {
+			return err
+		}
+	}
+
+	logger.Trace().Msgf("Current fingerprint: %v", sentryEvent.Fingerprint)
+	// Enhance message with object name
+	message := sentryEvent.Message
+	sentryEvent.Message = fmt.Sprintf("%s: %s", object.GetName(), sentryEvent.Message)
 
-	// Then, run kind-specific enhancers
-	switch objectRef.Kind {
-	case "Pod":
-		err = runPodEnhancer(ctx, objectRef, cachedObject, scope, sentryEvent)
+	// Adjust fingerprint.
+	// If there's already a non-empty fingerprint set, we assume that it was set by
+	// another enhancer, so we don't touch it.
+	if len(sentryEvent.Fingerprint) == 0 {
+		sentryEvent.Fingerprint = []string{message}
 	}
+	logger.Trace().Msgf("Fingerprint after adjustment: %v", sentryEvent.Fingerprint)
 
+	// Find the root owners and their corresponding object kinds
+	rootOwners, err := findRootOwners(ctx, &KindObjectPair{
+		kind:   kind,
+		object: object,
+	})
 	if err != nil {
-		logger.Error().Msgf("Error running an enhancer: %v", err)
+		return err
+	}
+
+	// Call the specific enhancer for the object
+	callObjectEnhancer(ctx, scope, &KindObjectPair{
+		kind,
+		object,
+	}, sentryEvent)
+
+	// Call specific enhancers for all root owners
+	// (there most likely is just one root owner)
+	for _, rootOwner := range rootOwners {
+		// We already called an enhancer for the object
+		// so we avoid calling it again
+		if rootOwner.object.GetName() == object.GetName() {
+			continue
+		}
+		callObjectEnhancer(ctx, scope, &rootOwner, sentryEvent)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type KindObjectPair struct {
+	kind   string
+	object metav1.Object
+}
+
+func findRootOwners(ctx context.Context, kindObjPair *KindObjectPair) ([]KindObjectPair, error) {
+
+	parents := kindObjPair.object.GetOwnerReferences()
+	// the owners slice to be returned
+	rootOwners := []KindObjectPair{}
+
+	// base case: the object has no parents
+	if len(parents) == 0 {
+		rootOwners = append(rootOwners, *kindObjPair)
```

Comment:
Q: why do we add the object itself if it's not managed by anything?

Reply:
Essentially I'm using DFS to find the objects are have no owningReferences. If an object is not managed by anything, then it is itself a root. I think this makes the base case cleaner.

In the case the enhancer calls an object with no owning references, I just make sure I don't call the enhancer twice on the same object by checking that the object and any object in its rootOwners slice are not equal.

Reply:
Maybe it makes sense from the DFS perspective, but interface-wise it doesn't sound clean to me. When I call the function called `findRootOwners`, I don't expect to have the object itself included in the output.

> I don't call the enhancer twice on the same object by checking that the object and any object in its rootOwners slice are not equal.

FWIW, checking by name only is not entirely safe (e.g. a pod theoretically can be owned by a resource with the same name).

Reply:
Thanks!